### PR TITLE
Topspin App5 loader v2

### PR DESCRIPTION
**Pull request by argerlt**
NOTE AT THE TOP: This might end up making more sense to either merge into #412, or to resubmit as a new PR with better contribution tracking. I'm open to discussion, as long as we get a working reader and everyone who worked on this gets credit. 

### Description of the change

This is a basic app5 reader that handles both kinds of common app5 formats produced by Nanomegas's Topspin software (exported versus copied directly from the file folder). It passes all unit tests, though I will admit the ones I've added are VERY basic and only have 91% coverage.

I don't believe it makes sense to include a writer, as Topspin populates its information with a very specific Metadata Layout based on the ProcedureName chosen, which also includes things like figure choices, background colors, formatting, etc. Without recreating all this information, the files do not open in TopSpin.

Additionally, another important point (and the reason I ended up making a new PR) is the layout of the Metadata structures as well as some of the data itself changes depending on the ProcedureName used and the version of Topspin. Thus, readers written for one file type won't necessarily work for all users everywhere.

After a lot of trial and error, I got this reader to work on the 50-odd .app5 files I have access to from across a few branches at NIST by ignoring the Metadata entirely unless absolutely necessary, and making a bit more generic of a dictionary builder. This will (hopefully?) make the reader a bit more future proof. 

As a bonus: app5 files often contain several (sometimes, hundreds) of 4Dstem datasets, which people may not want to load all at once, so i added a `dryrun` feature:

```
[In] >>> ]file_reader(file_from_export, dryrun=True)

[Out] >>> 
The following would be imported from /home/arg6/GitHub/rosettasciio/rsciio/tests/data/topspin/topspin_from_export.app5:
   - Test data B for rsciio, [3, 7, 29, 29]
      18d9446f-22bf-4fb1-8d13-338174e75d20/3526f008-a687-41fb-a21e-c21362241492
   - Test data B for rsciio, [11, 13]
      18d9446f-22bf-4fb1-8d13-338174e75d20/b0212ada-90b1-4117-8ed5-355717e4910b
   - Test data B for rsciio, [13, 17]
      18d9446f-22bf-4fb1-8d13-338174e75d20/b0e00442-4f83-4585-b84c-b82d2537b14b
   - Test data C for rsciio, [11, 13]
      b38446c6-6b25-4fee-9c62-6d8e4fcbeb5c/5a203dd8-ddbd-4775-8786-78d456e8b877
```

### Progress of the PR
- [x] Change implemented (can be split into several points),
- [x] update docstring (if appropriate),
- [x] update user guide (if appropriate),
- [x] add a changelog entry in the `upcoming_changes` folder (see [`upcoming_changes/README.rst`](https://github.com/hyperspy/rosettasciio/blob/main/upcoming_changes/README.rst)),
- [x] Check formatting of the changelog entry (and eventual user guide changes) in the `docs/readthedocs.org:rosettasciio` build of this PR (link in github checks)
- [x] add tests,
- [ ] ready for review.

### Minimal example of the bug fix or the new feature
```python
from rsciio.msa import file_reader
file_reader("your_msa_file.msa")
# Your new feature...
```
Note that this example can be useful to update the user guide.

### CODE REVIEW

**Comment by argerlt** [comment]
@ericpre and @maclariz  this is based on #412, and (I think?) properly credits all the commits from that branch. I think I'm going to slightly redo the git history and force-push an update to make credit clearer, but feel free to leave comments on the functions themselves, as those are basically done.

Also, @maclariz , I could not get the reader in #412 to work on any of my data, and to make sure I'm not creating the same problem in reverse for you, it would be helpful if I could get some of your datasets to turn into unittests stubs. I can reduce them down, I just need files created with Procedures other than "PED STEM"

File: rsciio/topspin/_api.py [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+        else:
+            # 2D composite image
+            axes = _get_2D_axes(metadata_string, name)
+            shape = h5_file[address].shape
+            axes[0]["size"] = shape[0]
+            axes[1]["size"] = shape[1]
+        datasets_list.append(
+            {
+                "axes": axes,
+                "metadata": meta_dict,
```
Comment (ericpre) [inline rsciio/topspin/_api.py]: `meta_dict` (all metadata parsed as python dictionary) go to `original_metadata`. `metadata` is for metadata defined in https://hyperspy.org/hyperspy-doc/current/reference/metadata.html or other extension metadata (https://docs.lumispy.org/en/stable/user_guide/metadata_structure.html, https://hyperspy.org/exspy/user_guide/metadata_structure.html, pyxem must be using some too, but I don't know where is documented).

File: rsciio/topspin/_api.py [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+    if dryrun:
+        filename = str(filename)
+        message = "The following would be imported from " + filename + ":"
+        for i, ds_dict in enumerate(datasets_list):
+            shape = [x["size"] for x in ds_dict["axes"]]
+            guid = guids_to_process[i][0]
+            name = ds_dict["metadata"]["Title"]
+            message += "\n   - {}, {}\n      {}".format(name, shape, guid)
+        print(message)
+        return ()
```
[the remark this answers is not in the document]
Comment (argerlt) [reply rsciio/topspin/_api.py]: I did try this function early on, but I don't think it makes sense for these files. Same is true for `read_metadata_from_file`. 

.app5 chose to store 4D datasets as groups with nx*ny sub-datasets. ie, a 1000x1000x256x256 datasets becomes a group containing 1,000,000 sub-groups, each with one 256-by-256 dataset and one string dataset. This makes a lot of sense from how TopSpin buffers data from the detector, but is massively frustrating from a data access standpoint.

The end effect though is any sensible hdf5 scanner will drill down to all 2-million subgroups, which takes almost as long as just loading the data. Even then, you still have to parse those text strings, which takes longer than this 3-line botique method. 

Similarly for the metadata reader, the "Metadata" in .app5 is really a custom XML format, which annoyingly means neither `rsciio.utils.hdf5.read_metadata_from_file` nor `rsciio.tools.XMLToDict` are useful. 

> Would it make sense to integrate this functionality in this function?
I don't think adding to this function makes sense.
1) , I'm looking for Groups, not datasets, and also only groups whose Datasets contain specific inormation.
2) The data I want to report is the shape of the scans, which isn't in the dataset shape and requires some XML parsing. This is what I'm getting from that `shape = [x["size"] for x in ds_dict["axes"]]` call, which has to access `_get_4D_axes` for 4D-STEM data.

But also, zooming out a bit, I think these functions are assuming a Dream3d NX -like structure, which is a sensible layout that a lot of files have, and .app5 is just a wacky outlier.

File: rsciio/topspin/_api.py [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+            axes[1]["size"] = shape[1]
+        datasets_list.append(
+            {
+                "axes": axes,
+                "metadata": meta_dict,
+                "original_metadata": metadata_string,
+            }
+        )
+
+    if dryrun:
```
[the remark this answers is not in the document]
Comment (argerlt) [reply rsciio/topspin/_api.py]: I dont think this is duplication of functionality. I wrote a bigger post above, but short version, I'm looking for a combination of Groups and data from those Groups, which cannot be grabbed with a normal, sensible hdf5 reader.

File: rsciio/topspin/_api.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ericpre) [reply rsciio/topspin/_api.py]: Okay, my comment is not relevant indeed - I didn't look into these details because, to be honest, I was expecting that the dataset will be stored in h5py dataset, etc.

[the remark this answers is not in the document]
Comment (argerlt) [reply rsciio/topspin/_api.py]: to be fair, when I told my boss "I think I can do this in 3 days", I thought the same thing.

**Comment by ericpre** [comment]
> I want to keep the artificial ones because they test 3 separate combinations of edge cases at two different user facilities, which has already caught a bunch of errors when writing this reader.

Yes, sure!

> Edge cases aside though, I fully agree with your points. How small is small enough? I found an .app5 that I can compress to 8.6 Mb file, but anything smaller will have to wait until the new year.

I would prefer to wait for the new files (as small as possible and suitable shape) along with good documentation of the hardware and software version. If this is convenient, this PR could be merged without the test file with the promise that they are added later one! ;)

**Comment by argerlt** [comment]
> It would be useful to document what hardware and software version are supported. 

Like list the supported versions in the file_reader docstring? Or is there a better way?

**Comment by argerlt** [comment]
@ericpre , this feature isn't done, but I think I've addressed all your concerns.

That said, between the black-vs-ruff, the repeated re-doing of the .app5 binaries, and a silly merge on my part, this  git history is messy. unless you are opposed, I'm going to open a new PR. 

Apologies, RosettaSciIo has some git control features I wasn't used to and I got a bit reckless with force pushing.

File: rsciio/topspin/_api.py [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+        if len(shape) == 2:
+            # this progress bar is really just a "1-of-1", but it still
+            # reports load time and confirms loading succeeded
+            for dummy_iter in tqdm(
+                [1],
+                desc="Loading {}: ".format(address),
+                disable=not show_progressbar,
+            ):
+                datasets_list[i]["data"] = np.asanyarray(h5_file[address])
+        elif len(shape) == 4:
```
Comment (ericpre) [inline rsciio/topspin/_api.py]: Do you have a file with a dataset with 3 dimension? Would it be possible to get one? If so, it can be done in a separate PR.

File: rsciio/tests/test_topspin.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+def test_with_hyperspy():
+    hs = pytest.importorskip("hyperspy.api", reason="hyperspy not installed")
+    for f in [file_A, file_B, file_C]:
+        out = file_reader(f, show_progressbar=False)
+        for dset in out:
+            signal = hs.signals.Signal2D(
+                data=dset["data"], axes=dset["axes"], metadata=dset["metadata"]
+            )
+            assert isinstance(signal, hs.signals.Signal2D)
```
Comment (ericpre) [inline rsciio/tests/test_topspin.py]: Can you check the value of the `scale`, `offsets` and `units` by using the value shown Topspin software? These are quite important and needs to be tested.

File: rsciio/topspin/_api.py [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+            # for 4D, data is loaded as ((x*y),ky,kz), then reshaped.
+            data = np.zeros([np.prod(shape[:2]), shape[2], shape[3]], dtype=np.uint16)
+            key_count = len(h5_file[address].keys())
+            for key in tqdm(
+                h5_file[address],
+                desc="Loading {}: ".format(address),
+                total=key_count,
+                disable=not show_progressbar,
+            ):
+                data[int(key)] = h5_file[address][key]["Data"]
```
Comment (ericpre) [inline rsciio/topspin/_api.py]: This may be non-negligeable overhead for open/closing each dataset. From a quick search on internet, it seems that this could be solved using the `h5pyd`, see:

https://github.com/HDFGroup/h5pyd/blob/f6376c33441911d4e6cba1e1ec1e36d51d548d69/docs/high/dataset.rst?plain=1#L439-L453

If you want to keep it as it in this PR, this is fine. I am hoping that this MultiManager interface will be useful for implementing efficient lazy loading.

If the dataset is contiguous which should be the default when chunking or compression is not used, another alternative is to load the dataset without `h5py` using the approach described in https://gist.github.com/maartenbreddels/09e1da79577151e5f7fec660c209f06e. This may be more brittle but there will be no overhead. As above, this should useful for efficient lazy loading.

File: rsciio/tests/data/topspin/topspin_test_data_README.txt [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+The following is a copy of the python code used to create small (200kb)
+app5 files from larger (approx 6GB) app5 files.
+
+The original files were collected at the Colorado School of Mines by
+NIST's Quantitative Nanostructure Characterization Group. The code below
+will not recreate the test data without the original files, but it is
+provided regarless:
+
+1) For transparency and validation.
+2) So that future users can potientially use it to make their own test data.
```
[the remark this answers is not in the document]
Comment (ericpre) [reply rsciio/tests/data/topspin/topspin_test_data_README.txt]: Having another look at this script - sorry, before without the python formatting, my brain was reluctant to read it properly! The issue with making this script more generic is the layout and I am also wondering if the following approach would work:
1. copy the whole file
2. open the newly copied file and reduce the size of the dataset in place, adjust other metadata accordingly
3. close the file!

This will have the following advantages:
- the layout will not change and the only thing that will be needed is to understand what needs to be changed to make the change in shape doesn't corrupt the file, but this should be already known considering that this file works!
- the characteristic/attribute of the dataset will not change, when creating a new dataset the dataset parameter may differ from the original one, for example if non-default parameter are used or if the default value of the parameter change across different hdf5/h5py versions.

If we can get this to work (more or less), I think that this would be very useful for adding support of files with different layout/file versions, because this script will help with troubleshooting - particularly since you already observed significant difference in the layout.

File: rsciio/topspin/_api.py [hunk 8]
```diff
@@ -0,0 +1,428 @@
...
+        Whether to show the progressbar or not. If True, shows progress
+        bar for each individual dataset loaded from the original file.
+    %s
+
+    Notes
+    -----
+    The hierarchy of the Metadata objects stored in app5 files changes
+    based on Topspin Proceedure and local microscope setup. RosettaSciIo
+    is still able to import datsets without this information, but be aware
+    that for new or customized setups some metadata entrees might be missing.
```
Comment (ericpre) [inline rsciio/topspin/_api.py]: In this case, the user should be made aware that some metadata are missing maybe by using the logger with level info or warning depending on criticality of this missing information? Document the behaviour here.

**Comment by ericpre** [comment]
Running co-pilot review to check for typos, etc. and maybe they will be some useful comments from co-pilot too!

File: rsciio/tests/test_topspin.py [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+#
+# You should have received a copy of the GNU General Public License
+# along with RosettaSciIO. If not, see <https://www.gnu.org/licenses/#GPL>.
+
+import xml.etree.cElementTree as ET
+from contextlib import redirect_stdout
+from io import StringIO
+from pathlib import Path
+
+import h5py
```
Comment (ericpre) [inline rsciio/tests/test_topspin.py]:
```suggestion
h5py = pytest.importorskip("h5py")
```
to be able to run the test suite when h5py is not installed, for example for the CI build with minimum dependencies.

File: rsciio/topspin/_api.py [hunk 10]
```diff
@@ -0,0 +1,428 @@
...
+import h5py
+import numpy as np
+from tqdm import tqdm
+
+from rsciio._docstrings import FILENAME_DOC, RETURNS_DOC
+
+
+def file_reader(filename, subset=None, dryrun=False, show_progressbar=True):
+    """
+    Read .app5 file format used by NanoMegas's Topspin software.
```
[the remark this answers is not in the document]
Comment (maclariz) [reply rsciio/topspin/_api.py]: @argerlt Basically, the app5 container supported loading multiple datasets into one file.  This was useful to us in the early days for exporting things from the microscope computer, loading them onto an offline analysis computer in Data Viewer, and then seeing the whole session.

Until we realised that if we want to unpack them and use our own Open Source tools, it's far better to do one dataset per app5...

Roughly speaking, NanoMegas have done things that work well if you are happy doing everything in their ecosystem, but not if you want to do things beyond that, or live view or process or anything like that.
